ros/src/computing/perception/localization/packages/kalman_filter_localizer/src/kalman_filter_localizer.cpp

```diff
@@ -0,0 +1,535 @@
+  X(IDX::Y) = initialpose.pose.pose.position.y;
+  X(IDX::YAW) = tf2::getYaw(initialpose.pose.pose.orientation);
+  X(IDX::YAWB) = 0.0;
+  X(IDX::VX) = 0.0;
+  X(IDX::WZ) = 0.0;
+
+  P(IDX::X, IDX::X) = initialpose.pose.covariance[0];
+  P(IDX::Y, IDX::Y) = initialpose.pose.covariance[6+1];
+  P(IDX::YAW, IDX::YAW) = initialpose.pose.covariance[6*5+5];
+  P(IDX::YAWB, IDX::YAWB) = 0.0001;
+  P(IDX::VX, IDX::VX) = 0.01;
+  P(IDX::WZ, IDX::WZ) = 0.01;
+
+  kf_.init(X, P, extend_state_step_);
+
+};
+
+/*
+ * callbackNDTPose
+ */
+void KalmanFilterNode::callbackNDTPose(const geometry_msgs::PoseStamped::ConstPtr &msg)
+{
+  current_ndt_pose_ptr_ = std::make_shared<geometry_msgs::PoseStamped>(*msg);
+};
+
+/*
+ * callbackTwist
+ */
+void KalmanFilterNode::callbackTwist(const geometry_msgs::TwistStamped::ConstPtr &msg)
+{
+  current_twist_ptr_ = std::make_shared<geometry_msgs::TwistStamped>(*msg);
+};
+
+/*
+ * initKalmanFilter
+ */
+void KalmanFilterNode::initKalmanFilter()
+{
+  Eigen::MatrixXd X = Eigen::MatrixXd::Zero(dim_x_, 1);
+  Eigen::MatrixXd P = Eigen::MatrixXd::Identity(dim_x_, dim_x_) * 1.0E3;
+  P(IDX::YAWB, IDX::YAWB) = cov_proc_yaw_bias_d_; // for yaw bias
+
+  kf_.init(X, P, extend_state_step_);
+}
+
+/*
+ * predictKinematicsModel
+ */
+void KalmanFilterNode::predictKinematicsModel()
+{
+  /*  == Nonlinear model ==
+   *
+   * x_{k+1}   = x_k + vx_k * cos(yaw_k + b_k) * dt
+   * y_{k+1}   = y_k + vx_k * sin(yaw_k + b_k) * dt
+   * yaw_{k+1} = yaw_k + (wz_k) * dt
+   * b_{k+1}   = b_k
+   * vx_{k+1}  = vz_k
+   * wz_{k+1}  = wz_k   
+   * 
+   * (b_k : yaw_bias_k)
+   */
+
+  /*  == Linearized model ==
+   *
+   * A = [ 1, 0, -vx*sin(yaw+b)*dt, -vx*sin(yaw+b)*dt, cos(yaw+b)*dt,  0]
+   *     [ 0, 1,  vx*cos(yaw+b)*dt,  vx*cos(yaw+b)*dt, sin(yaw+b)*dt,  0]
+   *     [ 0, 0,                 1,                 0,             0, dt]
+   *     [ 0, 0,                 0,                 1,             0,  0]
+   *     [ 0, 0,                 0,                 0,             1,  0]
+   *     [ 0, 0,                 0,                 0,             0,  1]
+   */
+
+  Eigen::MatrixXd X_curr(dim_x_, 1); // curent state
+  Eigen::MatrixXd X_next(dim_x_, 1); // predicted state
+  kf_.getCurrentX(X_curr);
+
+  Eigen::MatrixXd P_curr;
+  kf_.getCurrentP(P_curr);
+
+  const int d_dim_x = dim_x_ex_ - dim_x_;
+  const double yaw = X_curr(IDX::YAW);
+  const double yaw_bias = X_curr(IDX::YAWB);
+  const double vx = X_curr(IDX::VX);
+  const double wz = X_curr(IDX::WZ);
+  const double dt = kf_dt_;
+
+  /* Update for latest state */
+  X_next(IDX::X) = X_curr(IDX::X) + vx * cos(yaw + yaw_bias) * dt; // dx = v * cos(yaw)
+  X_next(IDX::Y) = X_curr(IDX::Y) + vx * sin(yaw + yaw_bias) * dt; // dy = v * sin(yaw)
+  X_next(IDX::YAW) = X_curr(IDX::YAW) + (wz)*dt;                   // dyaw = omega + omega_bias
+  X_next(IDX::YAWB) = yaw_bias;
+  X_next(IDX::VX) = vx;
+  X_next(IDX::WZ) = wz;
+
+  while (std::fabs(X_next(IDX::YAW)) > M_PI)
```

Comment:
I think that it is better to convert to a vector once and normalize it by atan2
https://github.com/CPFL/Autoware/blob/ae86b06f7638b8b3b4a12381a2844d26ea569f3f/ros/src/common/libs/amathutils_lib/src/vector2d.cpp#L15
https://github.com/CPFL/Autoware/blob/ae86b06f7638b8b3b4a12381a2844d26ea569f3f/ros/src/common/libs/amathutils_lib/src/vector2d.cpp#L30

Reply:
I've modified following your suggestion.